parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java

```diff
@@ -58,40 +58,43 @@ public void poll(Consumer<PollContext<K, V>> usersVoidConsumptionFunction) {
     @Override
     @SneakyThrows
     public void pollAndProduceMany(Function<PollContext<K, V>, List<ProducerRecord<K, V>>> userFunction,
-        Consumer<ConsumeProduceResult<K, V, K, V>> callback) {
-        // todo refactor out the producer system to a sub class
+                                   Consumer<ConsumeProduceResult<K, V, K, V>> callback) {
         if (!getOptions().isProducerSupplied()) {
             throw new IllegalArgumentException("To use the produce flows you must supply a Producer in the options");
         }
 
         // wrap user func to add produce function
-        Function<PollContextInternal<K, V>, List<ConsumeProduceResult<K, V, K, V>>> wrappedUserFunc = context -> {
-            List<ProducerRecord<K, V>> recordListToProduce = carefullyRun(userFunction, context.getPollContext());
+        Function<PollContextInternal<K, V>, List<ConsumeProduceResult<K, V, K, V>>> wrappedUserFunc =
+                context -> userFunctionWrap(userFunction, context);
 
-            if (recordListToProduce.isEmpty()) {
-                log.debug("No result returned from function to send.");
-            }
-            log.trace("asyncPoll and Stream - Consumed a record ({}), and returning a derivative result record to be produced: {}", context, recordListToProduce);
+        supervisorLoop(wrappedUserFunc, callback);
+    }
+
+    private List<ConsumeProduceResult<K, V, K, V>> userFunctionWrap(Function<PollContext<K, V>, List<ProducerRecord<K, V>>> userFunction,
```

Comment:
nitpick - better name for the function? producingWrapper? something to tie in with the fact that it is the part that does produce after userFunction processed consumed messages? 
as it stands it is not really telling me anything about what that function does.

Reply:
May be something along the lines of 
```
Function<PollContextInternal<K, V>, List<ConsumeProduceResult<K, V, K, V>>> producingUserFunctionWrapper = context -> {
    List<ProducerRecord<K, V>> recordListToProduce = carefullyRun(userFunction, context.getPollContext());
    return produceMessages(recordListToProduce, context));
    };
```
or just renaming maybe
```
Function<PollContextInternal<K, V>, List<ConsumeProduceResult<K, V, K, V>>> producingUserFunctionWrapper = context -> processAndProduceResults(userFunction, context);
```

Reply:
slack discussion